# Topologically associated domains (TADs) analysis

library(rtracklayer)
library(mosaic)

# This script assumes that the script HiC_V*.R has been run before

# File Neu.50000_blocks.bed from Rajarajan et al. (2018)
TADS50 <- import(file.choose(), format="bed")
<<<<<<< HEAD
TADS50.bis <- import(file.choose(), format = "bed")
=======
# File Neu(1).100000_blocks.bed from Rajarajan et al. (2018)
>>>>>>> efcc967b6605c7f4ef5c44bcc184fc817b0d3551
TADS100 <- import(file.choose(), format="bed")
TADS10 <- import(file.choose(), format="bed")
#The file generated by Arrowhead is a bedpe file
TADs5.bedpe <- import(file.choose(),format="bedpe")

<<<<<<< HEAD
#To facilitate working manipulations we've converted bedpe file into csv file
df.TADs5 <- read.csv2(file.choose(), sep="\t",header=FALSE)
=======
#To facilitate working manipulations we've converted bedpe file into tab-delimited file
# Warning! the file TADs_5000_all_blocks has a header line contrary to other tab-delimited files 
  df.TADs5 <- read.table(file.choose(), sep="\t")
>>>>>>> efcc967b6605c7f4ef5c44bcc184fc817b0d3551
TADs5.bed <- df.TADs5[,c(1,2,3)]
colnames(TADs5.bed) <- c("chr", "start", "end")

df.TADs10 <- read.table(file.choose(), sep="\t",header=T)
TADs10.bed <- df.TADs10[,c(1,2,3)]
colnames(TADs10.bed) <- c("chr", "start", "end")

#For a better retrieving of TADs, we make two kind of index: 
  # - a global index
  # - a index per chr
#The main aim of indexing is to find a relation between nearby TADs 

TADs5.bed <- TADs5.bed[with(TADs5.bed, order(chr,start)),]
TADs5.bed$index.TADs <- seq(1, nrow(TADs5.bed), 1)

for (i in unique(TADs5.bed$chr)) {
  n <- rownames(TADs5.bed[TADs5.bed$chr == i,])
  
  TADs5.bed[n,"index.TADs.chr"] <- seq(1, length(n), 1)
}

TADS5 <- GRanges(TADs5.bed)
mcols(TADS5)$index.TADs <- TADs5.bed$index.TADs
mcols(TADS5)$index.TADs.chr <- TADs5.bed$index.TADs.chr

TADs10.bed <- TADs10.bed[with(TADs10.bed, order(chr,start)),]
TADs10.bed$index.TADs <- seq(1, nrow(TADs10.bed), 1)

for (i in unique(TADs10.bed$chr)) {
  n <- rownames(TADs10.bed[TADs10.bed$chr == i,])
  
  TADs10.bed[n,"index.TADs.chr"] <- seq(1, length(n), 1)
}

mcols(TADS10)$index.TADs <- TADs10.bed$index.TADs
mcols(TADS10)$index.TADs.chr <- TADs10.bed$index.TADs.chr


#Number of TADs in the two files
length(TADS50)
#[1] 2525
length(TADS50.bis)
#[1] 2525
length(TADS10)
#[1] 4319
length(TADS100)
#[1] 1248
length(TADS5)
#1991

# Much fewer TADs of 100kb blocks
sum(width(reduce(TADS50)))
#[1] 2 316 700 000

sum(width(reduce(TADS50.bis)))
#[1] 2 315 450 000

sum(width(reduce(TADS50.bis))) / sum(width(reduce(TADS50)))
# 99% of author file is preserved with our own methodology

sum(width(reduce(TADS10)))
#[1] 1 608 620 000

sum(width(reduce(TADS100)))
#[1] 2 205 200 000

sum(width(reduce(TADS5)))
#[1] 845 841 912

#Some descriptive statistics on 5Kb resolution 
#Number of TADs per chr
table(TADs5.bed$chr)
TADs5.bed$width <- TADs5.bed$end - TADs5.bed$start

hist(TADs5.bed$width)
summary(TADs5.bed$width)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#60000  135000  380000  444146 565000 2660000 
#The most high part of TADs are little domains

#Mean coverage of TADs per chr
aggregate(TADs5.bed$width~TADs5.bed$chr,FUN = mean)
summary(aggregate(TADs5.bed$width~TADs5.bed$chr,FUN = mean)[,2])
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#263317  401936  463812  454474  492671  705179 

# Genomic coverage is about the same for 50kb and 100kb blocks
# Creating a GRanges object with the Epigenomics roadmap-based clusters
# Extract chromosome for each cluster in 1-22,X format from chr1-chr22,chrX format
chr.cluster = tapply(genes.enhancers.EGRM$chr,compo.epg$membership[genes.enhancers.EGRM$geneSymbol],function(vec) substring(vec[1],4))
cluster.GRanges = GRanges(seqnames = chr.cluster,ranges = IRanges(start.cluster,end.cluster))

#For each cluster,we fit the cluster id
names(cluster.GRanges) <- names(longueur.cluster)
GRange.genes.enhancers <- GRanges(seqnames = genes.enhancers.EGRM$chr, ranges = IRanges(genes.enhancers.EGRM$enhancerstart, genes.enhancers.EGRM$enhancerstop), geneSymbol = genes.enhancers.EGRM$geneSymbol, TSS_TES = IRanges(genes.enhancers.EGRM$TSS, genes.enhancers.EGRM$TES))


# Number of TADs overlapped by each cluster
cluster.TADs50 <- findOverlaps(cluster.GRanges, TADS50)
table(countLnodeHits(cluster.TADs50))
#0   1   2   3   4   5   6 
#173 373 218  83  25   8   1 
round(tally(countLnodeHits(cluster.TADs50),format="percent"))
#0  1  2  3  4  5  6 
#20 42 25  9  3  1  0 

cluster.TADs10 <- findOverlaps(cluster.GRanges, TADS10)
table(countLnodeHits(cluster.TADs10))
# 0   1   2   3   4   5   6   9  16 
#77 459 236  81  15   9   2   1   1 
round(tally(countLnodeHits(cluster.TADs10),format="percent"))
#0  1  2  3  4  5  6  9 16 
#9 52 27  9  2  1  0  0  0 


cluster.TADs100 <- findOverlaps(cluster.GRanges, TADS100)
table(countLnodeHits(cluster.TADs100))
#0   1   2   3   4   5 
#240 349 202  68  18   4 
round(tally(countLnodeHits(cluster.TADs100),format="percent"))
#0  1  2  3  4  5 
#27 40 23  8  2  0 

cluster.TADs5 <- findOverlaps(cluster.GRanges, TADS5)
#For each overlapped TADs we have the cluster id
mcols(TADS5)[subjectHits(cluster.TADs5), "n.cluster"] <- queryHits(cluster.TADs5)


subset(TADS5,n.cluster == 689); genes.enhancers.EGRM[genes.enhancers.EGRM$geneSymbol %in% names(compo.epg$membership[compo.epg$membership == 689]),c("geneSymbol", "enhancerstart", "enhancerstop", "TSS", "TES")]
table(countLnodeHits(cluster.TADs5))
#0   1   2   3   4   7 
#293 486  77  19   5   1 
round(tally(countLnodeHits(cluster.TADs5),format="percent"))
#0  1  2  3  4  7 
#33 55  9  2  1  0 

#What is the proportion of clusters which are completly inside TADs? 
n.inside.clusters <- sum(start(cluster.GRanges[queryHits(cluster.TADs5)]) >= start(TADS5[subjectHits(cluster.TADs5)]) & end(cluster.GRanges[queryHits(cluster.TADs5)]) <= end(TADS5[subjectHits(cluster.TADs5)]))
#[1] 143
n.inside.clusters / length(cluster.GRanges)
#[1] 0.1623156

TADoverlapAnalysis = function(cluster.TADs,TADs)
{
  #Clusters which overlap several TADs
  #Determines the structure of overlapped TADs by cluster
  #and tallies number of TADs involved in each structure 
  names.several.TADs <- names(table(queryHits(cluster.TADs))[table(queryHits(cluster.TADs))!=1])
clusters.several.TADs <- cluster.TADs[queryHits(cluster.TADs) %in% names.several.TADs]

distinct.TADs <- 0
closer.distinct.TADs <- 0 
nested.TADs <- 0 
overlaps.TADs <- 0

for(i in unique(queryHits(clusters.several.TADs))){
  tmp.subjectHits <- subjectHits(clusters.several.TADs[queryHits(clusters.several.TADs) == i])
  tmp.TADs <- TADs[tmp.subjectHits]
  
  n <- length(tmp.TADs)
  
  unique.span <- sum(width(reduce(tmp.TADs)))
  tt.span <- sum(width(tmp.TADs))
  w <- width(tmp.TADs[which.max(width(tmp.TADs))])
  
  # If the genome length uniquely spanned by the TADs equals the 
  # sum of the TADs lengths then all TADs are distincts
  if(unique.span == tt.span){
    distinct.TADs =  distinct.TADs + n
    # If last TAD for this cluster is n-1 positions from first TAD in global index
    # then all TADs are consecutive
    if(mcols(tmp.TADs)$index.TADs[n] == mcols(tmp.TADs)$index.TADs[1] + n -1){
      closer.distinct.TADs = closer.distinct.TADs +n 
    }
  }
  # Else if the genome length uniquely spanned by the TADs equals the
  # width of the largest TAD, then all other TADs are nested in that largest TAD
  else if(unique.span == w){
    nested.TADs = nested.TADs + n
  }
  # Else we conclude that we have some other form of overlap
  else{
    overlaps.TADs = overlaps.TADs + n
  }

}

list(distinct.TADs=distinct.TADs,closer.distinct.TADs=closer.distinct.TADs,overlaps.TADs=overlaps.TADs,nested.TADs=nested.TADs)
}

clusterTADoverlapAnalysis = function(cluster.TADs,TADs)
{
  #Clusters which overlap several TADs
  #Determines the structure of overlapped TADs by cluster
  #and tallies number of clusters with each structure 
  names.several.TADs <- names(table(queryHits(cluster.TADs))[table(queryHits(cluster.TADs))!=1])
  clusters.several.TADs <- cluster.TADs[queryHits(cluster.TADs) %in% names.several.TADs]
  
  distinct.TADs <- 0
  closer.distinct.TADs <- 0 
  nested.TADs <- 0 
  overlaps.TADs <- 0
  
  for(i in unique(queryHits(clusters.several.TADs))){
    tmp.subjectHits <- subjectHits(clusters.several.TADs[queryHits(clusters.several.TADs) == i])
    tmp.TADs <- TADs[tmp.subjectHits]
    
    n <- length(tmp.TADs)
    
    unique.span <- sum(width(reduce(tmp.TADs)))
    tt.span <- sum(width(tmp.TADs))
    w <- width(tmp.TADs[which.max(width(tmp.TADs))])
    
    # If the genome length uniquely spanned by the TADs equals the 
    # sum of the TADs lengths then all TADs are distincts
    if(unique.span == tt.span){
      distinct.TADs =  distinct.TADs + 1
      # If last TAD for this cluster is n-1 positions from first TAD in global index
      # then all TADs are consecutive
      if(mcols(tmp.TADs)$index.TADs[n] == mcols(tmp.TADs)$index.TADs[1] + n -1){
        closer.distinct.TADs = closer.distinct.TADs +1 
      }
    }
    # Else if the genome length uniquely spanned by the TADs equals the
    # width of the largest TAD, then all other TADs are nested in that largest TAD
    else if(unique.span == w){
      nested.TADs = nested.TADs + 1
    }
    # Else we conclude that we have some other form of overlap
    else{
      overlaps.TADs = overlaps.TADs + 1
    }
    
  }
  
  list(cluster.distinct.TADs=distinct.TADs,cluster.closer.distinct.TADs=closer.distinct.TADs,cluster.overlaps.TADs=overlaps.TADs,cluster.nested.TADs=nested.TADs)
}

TADs5.overlap = TADoverlapAnalysis(cluster.TADs5,TADS5)

#161 clusters which are overlapped distinct TADs
#32 clusters which overlapped overlapped TADs
#45 clusters which overlapped nested TADs
TADs5.overlap$closer.distinct.TADs == TADs5.overlap$distinct.TADs 
#[1] TRUE
#100% of distinct TADs are the closest

(TADs5.overlap$distinct.TADs + TADs5.overlap$overlaps.TADs + TADs5.overlap$nested.TADs) == length(clusters.several.TADs)
#[1] TRUE


TADs10.overlap = TADoverlapAnalysis(cluster.TADs10,TADS10)
TADs10.overlap
#$distinct.TADs
#[1] 479

#$closer.distinct.TADS
#[1] 8

#$overlaps.TADs
#[1] 173

#$nested.TADs
#[1] 205

cluster.TADs10.overlap = clusterTADoverlapAnalysis(cluster.TADs10,TADS10)
cluster.TADs10.overlap
#$cluster.distinct.TADs
#[1] 207

#$cluster.closer.distinct.TADs
#[1] 4

#$cluster.overlaps.TADs
#[1] 43

#$clusters.nested.TADs
#[1] 95

(cluster.TADs10.overlap$cluster.distinct.TADs + cluster.TADs10.overlap$cluster.overlaps.TADs + cluster.TADs10.overlap$cluster.nested.TADs) == sum(countLnodeHits(cluster.TADs10)>1)

# Proportion of clusters touching at least one TAD that touch at least 2 distinct TADs
# lower bound 
cluster.TADs10.overlap$cluster.distinct.TADs/(length(cluster.GRanges)-sum(countLnodeHits(cluster.TADs10)==0))
#[1] 0.2574627

# upper bound including clusters overlapping multiple TADs that are not all distinct
(cluster.TADs10.overlap$cluster.distinct.TADs+cluster.TADs10.overlap$cluster.overlaps.TADs)/(length(cluster.GRanges)-sum(countLnodeHits(cluster.TADs10)==0))
#[1] 0.3109453

#The idea behind a cluster which overlapped several different TADs is the membership of this latters to a big meta-TADs
#https://dumas.ccsd.cnrs.fr/dumas-01628629/document (22-23)

#We possibly 3 correlation levels for TADs: -weak(distinct TADs), -medium(overlapped TADs) and -high (nested TADs)
#each level correspond to the structure of the belonging of clusters to TADs


# Gene-enhancer Epigenomics Roadmap pairs overlap with TADs
vec.X = ifelse(genes.enhancers.EGRM$TSS<=genes.enhancers.EGRM$enhancerstop,genes.enhancers.EGRM$TSS, genes.enhancers.EGRM$enhancerstart)
vec.Y = ifelse(genes.enhancers.EGRM$TSS>genes.enhancers.EGRM$enhancerstop,genes.enhancers.EGRM$TES, genes.enhancers.EGRM$enhancerstop)
genes.enhancers.EGRM.GRanges=GRanges(substring(genes.enhancers.EGRM$chr,4),ranges = IRanges(vec.X,vec.Y))

# Number of TADs overlapped by each gene-enhancer pair
genes.enhancers.EGRM.TADs50 <- findOverlaps(genes.enhancers.EGRM.GRanges, TADS50)
gecTADs50 = countLnodeHits(genes.enhancers.EGRM.TADs50)
table(gecTADs50)
#0   1   2   3   4   5   6 
#256 560 327 134  29   8   1 
round(tally(gecTADs50,format="percent"))
#0  1  2  3  4  5  6 
#19 43 25 10  2  1  0 
round(tally(gecTADs50[gecTADs50>0],format="percent"))
#1  2  3  4  5  6 
#53 31 13  3  1  0

genes.enhancers.EGRM.TADs50.bis <- findOverlaps(genes.enhancers.EGRM.GRanges, TADS50.bis)
gecTADs50.bis = countLnodeHits(genes.enhancers.EGRM.TADs50.bis)
table(gecTADs50.bis)
#0   1   2   3   4   5   6 
#256 566 317 131  33  11   1 
round(tally(gecTADs50.bis,format="percent"))
#0  1  2  3  4  5  6 
#19 43 24 10  3  1  0 
round(tally(gecTADs50.bis[gecTADs50.bis>0],format="percent"))
#1  2  3  4  5  6 
#53 30 12  3  1  0 


genes.enhancers.EGRM.TADs100 <- findOverlaps(genes.enhancers.EGRM.GRanges, TADS100)
gecTADs100 = countLnodeHits(genes.enhancers.EGRM.TADs100)
table(gecTADs100)
#0   1   2   3   4   5 
#363 520 291 111  25   5 
round(tally(gecTADs100,format="percent"))
#0  1  2  3  4  5 
#28 40 22  8  2  0 
round(tally(gecTADs100[gecTADs100>0],format="percent"))
#1  2  3  4  5 
#55 31 12  3  1 

genes.enhancers.EGRM.TADs5 <- findOverlaps(genes.enhancers.EGRM.GRanges, TADS5)
gecTADs5 = countLnodeHits(genes.enhancers.EGRM.TADs5)
table(gecTADs5)
# 0   1   2   3   4   7 
#458 719 116  15   5   2 
round(tally(gecTADs5,format="percent"))
#0  1  2  3  4  7 
#35 55  9  1  0  0 
round(tally(gecTADs5[gecTADs5>0],format="percent"))
#1  2  3  4  7 
#84 14  2  1  0
#17% of enhancer-promoter pairs span 2 or more TADs, which is the same as results seens in other tissues

genes.enhancers.EGRM.TADs10 <- findOverlaps(genes.enhancers.EGRM.GRanges, TADS10)
gecTADs10 = countLnodeHits(genes.enhancers.EGRM.TADs10)
table(gecTADs10)
# 0   1   2   3   4   5   6   9  16 
#146 711 330  97  16  10   2   1   2 
round(tally(gecTADs10,format="percent"))
#0  1  2  3  4  5  6  9 16 
#11 54 25  7  1  1  0  0  0 
round(tally(gecTADs10[gecTADs10>0],format="percent"))
#1  2  3  4  5  6  9 16 
#61 28  8  1  1  0  0  0

# Eliminating overlapping TADs
keep = numeric(0)
for (i in 1:length(TADS10))
{
  if(!any(seqnames(TADS10[i])==seqnames(TADS10)&((start(TADS10[i])<=start(TADS10)&end(TADS10[i])>end(TADS10))|(start(TADS10[i])<start(TADS10)&end(TADS10[i])>=end(TADS10)))))
    keep = c(keep,i)
}
TADs10.distinct = TADS10[keep]

cluster.TADs10.distinct <- findOverlaps(cluster.GRanges, TADs10.distinct)
table(countLnodeHits(cluster.TADs10.distinct))
#0   1   2   3   4   5   6  14 
#84 546 182  52  12   2   2   1 
round(tally(countLnodeHits(cluster.TADs10.distinct),format="percent"))
#0  1  2  3  4  5  6 14 
#10 62 21  6  1  0  0  0 

sum(countLnodeHits(cluster.TADs10.distinct)>1)/(length(cluster.GRanges)-sum(countLnodeHits(cluster.TADs10.distinct)==0))
#[1] 0.314931
sum(countLnodeHits(cluster.TADs10.distinct)>1)/(length(cluster.GRanges)-sum(countLnodeHits(cluster.TADs10)==0))
#[1] 0.3121891
# Close to the upper bound of the clusterTADoverlapAnalysis function


# Number of distinct TADs overlapped by each gene-enhancer pair
genes.enhancers.EGRM.TADs10.distinct <- findOverlaps(genes.enhancers.EGRM.GRanges, TADs10.distinct)
gecTADs10.distinct = countLnodeHits(genes.enhancers.EGRM.TADs10.distinct)
table(gecTADs10.distinct)
#0   1   2   3   4   5   6  14 
#157 832 253  56  11   2   2   2 
round(tally(gecTADs10.distinct,format="percent"))
#0  1  2  3  4  5  6 14 
#12 63 19  4  1  0  0  0 
round(tally(gecTADs10.distinct[gecTADs10.distinct>0],format="percent"))
#1  2  3  4  5  6 14 
#72 22  5  1  0  0  0 



